File: pb/src/main/java/org/sqlbuilder/pb/objects/Arg.java
package org.sqlbuilder.pb.objects;

import org.sqlbuilder.common.*;
import org.sqlbuilder.pb.PbNormalizer;

import java.util.Comparator;

public class Arg implements HasId, Insertable, Comparable<Arg>
{
	private static final Comparator<Arg> COMPARATOR = Comparator //
			.comparing(Arg::getExample) //
			.thenComparing(Arg::getText) //
			.thenComparing(Arg::getN) //
			.thenComparing(Arg::getF, Comparator.nullsFirst(Comparator.naturalOrder()));

	public static final SetCollector<Arg> COLLECTOR = new SetCollector<>(COMPARATOR);

	private final Example example;

	private final String text;

	@NotNull
	private final ArgN n;

	@Nullable
	private final Func f;

	// C O N S T R U C T O R

	public static Arg make(final Example example, final String text, @NotNull final String n, final String f)
	{
		var a = new Arg(example, text, n, f);
		COLLECTOR.add(a);
		return a;
	}

	private Arg(final Example example, final String text, final String n, final String f)
	{
		assert n != null && !n.isEmpty();
		this.example = example;
		this.text = PbNormalizer.normalize(text);
		this.n = ArgN.make(n);
		this.f = f == null || f.isEmpty() ? null : Func.make(f.toLowerCase());
	}

	// A C C E S S

	public Example getExample()
	{
		return this.example;
	}

	public String getText()
	{
		return this.text;
	}

	@Nullable
	public Func getF()
	{
		return this.f;
	}

	@NotNull
	public ArgN getN()
	{
		return this.n;
	}

	@Override
	public Integer getIntId()
	{
		return COLLECTOR.get(this);
	}

	// O R D E R

	@Override
	public int compareTo(@NotNull final Arg that)
	{
		return COMPARATOR.compare(this, that);
	}

	// T O S T R I N G

	@Override
	public String toString()
	{
		return String.format("arg %s[%s][%s]", example, n, f);
	}

	// I N S E R T

	@RequiresIdFrom(type = Func.class)
	@RequiresIdFrom(type = Example.class)
	@Override
	public String dataRow()
	{
		// (argid),text,n,f,exampleid
		return String.format("'%s','%s',%s,%s", //
				Utils.escape(text), //
				n.getArgn(), //
				Func.getIntId(f), //
				example.getIntId());
	}

	@Override
	public String comment()
	{
		return String.format("%s,%s", n.getArgn(), f == null ? null : f.getFunc());
	}
}

File: pb/src/main/java/org/sqlbuilder/pb/objects/Arg.kt
package org.sqlbuilder.pb.objects

import org.sqlbuilder.annotations.RequiresIdFrom
import org.sqlbuilder.common.*
import org.sqlbuilder.pb.PbNormalizer

class Arg private constructor(example: Example, text: String, val type: String) : HasId, Insertable, Comparable<Arg> {

    private val example: Example

    private val text: String

    private val n: ArgType

    private val f: Func?

    // C O N S T R U C T O R

    init {
        assert(!type.isEmpty())
        this.example = example
        this.text = PbNormalizer.normalize(text)
        val fn = extractFN(type)
        n = fn.first
        f = fn.second
    }

    private fun extractFN(type: String): Pair<ArgType, Func?> {
        val fields = type.split("-")
        val nFields = fields.size
        // find first field starting with 'ARG'
        var index = fields.indexOfFirst { it.startsWith("ARG") }
        // make
        val n = ArgType.make(fields[index].replace("ARG", ""))
        val f = if (nFields > index + 1) Func.makeOrNull(fields[index + 1]) else null
        return n to f
    }

    // N I D

    override fun getIntId(): Int {
        return COLLECTOR.apply(this)
    }

    // O R D E R

    override fun compareTo(that: Arg): Int {
        return COMPARATOR.compare(this, that)
    }

    // T O S T R I N G

    override fun toString(): String {
        return "arg $example[$n][$f]"
    }

    // I N S E R T

    @RequiresIdFrom(type = Func::class)
    @RequiresIdFrom(type = Example::class)
    override fun dataRow(): String {
        return "'${Utils.escape(text)}','${n.argType}',${if (f == null) "NULL" else Func.getIntId(f)},${example.intId}"
    }

    override fun comment(): String {
        return "${n.argType},${f?.func}"
    }

    companion object {

        private val COMPARATOR: Comparator<Arg> = Comparator
            .comparing<Arg, Example> { it.example }
            .thenComparing<String> { it.text }
            .thenComparing<ArgType> { it.n }
            .thenComparing<Func>({ it.f }, Comparator.nullsFirst<Func>(Comparator.naturalOrder()))

        @JvmField
        val COLLECTOR = SetCollector<Arg>(COMPARATOR)

        fun make(example: Example, text: String, type: String): Arg {
            val a = Arg(example, text, type)
            COLLECTOR.add(a)
            return a
        }
    }
}
